### ISSUE TITLE
pytest.ini file in tests folder prevents single-test UI status update

### ISSUE THREAD

**Issue by CarloDePieri** [kondratyev-nv/vscode-python-test-adapter#198]
littlefoxteam.vscode-python-test-adapter: v0.6.1
vscode version: 1.50.0
Commit: 93c2f0fbf16c5a4b10e4d5f89737d9c2c25488a3
Date: 2020-10-07T06:01:33.073Z
Electron: 9.2.1
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Linux x64 5.8.2-arch1-1-vfio

My project structure is currently:

```bash
.
└── my_python_module/
    └──tests/
        ├── pytest.ini
        └── test_sometest.py
```

This is a valid project structure described in the [pytest docs](https://docs.pytest.org/en/stable/customize.html) and that works without problems when executing `pytest tests` from the project root folder.

Both the default Python extension's Test View and Test Explorer find all tests without any issue and both execute them successfully when clicking on the 'run all' buttons.

The default Test View allows to run a single test by clicking on the relative button or on the lens in the code, which executes the test and updates the Test View ui and editor linting accordingly. 

Trying to execute a single test (or a single class, or a single file) using Test Explorer actually runs the selected tests (and this can be verified by the python test adapter log or by reviewing the xml output file or by having some side effects in a test) BUT the UI in the Test Explorer view does not get updated nor the editor gutter or error decorations change.

This seems caused by the presence of `pytest.ini` in the `tests` folder because when moving it into the project root folder Test Explorer returns behaving as expected.

To reproduce this, I've prepared a [minimal demo repo](https://github.com/CarloDePieri/pytest-vscode-python-test-adapter-issue):
1. Clone the repo
2. Install pytest if needed (Pipfile is included)
3. Try to launch a single test in vscode: the ui won't update
4. Move `pytest.ini` in the project root folder or simply delete it
5. Try again to launch a single test: it will work now

I've got no idea if this is the intended behavior, but since I could not find all of this documented anywhere I thought to report it. Thank you for you hard work on the extension!

**Comment by Makishima** [comment]
It's affected me too. Not sure that it is work as intended.

**Comment by kondratyev-nv** [comment]
@CarloDePieri @Makishima Thank you for the issue! It looks like `--rootdir` has to be set to a `python.testing.cwd` always. In this case, execution will be consistent between the extension and pytest. I'll make a fix.

**Comment by kondratyev-nv** [comment]
@CarloDePieri @Makishima I've published a new version 0.6.2 that always sets `--rootdir` argument for pytest to a value of `python.testing.cwd` (default is workspace folder). Please, update and let me know if the issue persists.

**Comment by Makishima** [comment]
@kondratyev-nv Yep, it's worked. 
Is there a way to discover and show tests in this setup?

```
1. root dir/
    1.1. project 1/
        1.1.1. some.py
        1.1.2. pytest.ini
	1.2. project 2/
		1.2.1. some.py
		1.2.2. pytest.ini
```

**Comment by kondratyev-nv** [comment]
@Makishima So discovery is not working for you in this setup? Does it work when you run pytest from the `root dir`?

**Comment by Makishima** [comment]
I investigated a bit and this more pytest behavior than an extension. If there no `pytest.ini` in `root dir` so `pytest` will discover with default file names like `_test.py`
___
Example 1.
`pytest --collect-only` in
```
.
├── p1
│   ├── p1t_test.py
│   └── pytest.ini
└── p2
    ├── p2t_test.py
    └── pytest.ini
```
with contest of `pytest.ini`
```
[pytest]
python_files = p2t_*.py
python_functions = *_p2
```
will discover all tests (by default `pytest` looks for `*_test.py` and other wildcard masks)
```
<Module p1/p1t_test.py>
  <Function test_p1>
<Module p2/p2t_test.py>
  <Function test_p2>
```

Example 2.
If I override the default settings with `pytest.ini` in `root dir`
```
.
├── p1
│   ├── p1t_test.py
│   └── pytest.ini
├── p2
│   ├── p2t_test.py
│   └── pytest.ini
└── pytest.ini
```
It will found the only p2 test
```
<Module p2/p2t_test.py>
  <Function test_p2>
```

Example 3.
And finally if I run `pytest p1 --collect-only` or `pytest p2 --collect-only` it will discover test depend on directory `pytest.ini`
```
pytest p1 --collect-only

rootdir: /home/nikolay/ptest/p1, configfile: pytest.ini
collected 0 items
```
```
pytest p2 --collect-only

rootdir: /home/nikolay/ptest/p2, configfile: pytest.ini
collected 1 item

<Module p2t_test.py>
  <Function test_p2>
```
____

So the question is are extension is able to run discovery separatly for each directory in workspace? Like `pytest <DIR_NAME> --collect-only`

**Comment by kondratyev-nv** [comment]
@Makishima Thank you for the clarification. Yeah, in general, the extension is trying to behave similarly to pytest. I believe this is expected for most of the users.
So, as I understand you have multiple directories with pytest.ini files and run your tests by calling `pytest <specific directory>`. The extension is working with a single folder at a time (`python.testing.cwd`, top-level by default). So discovering and running tests is the same as running `pytest` from the top-level folder. I'm not sure how to achieve what you want from the top-level, as pytest itself is not able to do that. You can ask pytest team if there is a way to correctly run tests with multiple .ini files - https://github.com/pytest-dev/pytest/. 
I can suggest taking a look into writing top-level `conftest.py` which will hande the logic of what tests should be executed. Also, I've tried running `pytest p1 p2 --collect-only`, but it does not seem to work, because of [rootdir detection rules](https://docs.pytest.org/en/stable/customize.html#finding-the-rootdir). From the perspective of the extension, I can suggest creating a [workspace](https://code.visualstudio.com/docs/editor/multi-root-workspaces) and adding p1 and p2 as separate workspace folders. I'm not sure it will work for you if you have any top-level files you're working with (It will also require duplicating .vscode settings to each folder). And of course, you can run tests in each folder at a time by changing `python.testing.cwd` or `python.testing.pytestArgs`.

**Comment by CarloDePieri** [comment]
@kondratyev-nv the issue is solved with 0.6.2, thanks for the quick fix!

@Makishima The issue there is that a pytest process can have only one `Conftest.inipath` object, so it can only parse one ini file at a time. To make your configuration work you could have a single `pytest.ini` at the root level like this one:

```
[pytest]
testpaths =
    p1
    p2
python_files = p2t_*.py p1t_*.py
python_functions = *_p2 *_p1
```
With this, both pytest (launched with no arguments from the root dir) and the Test Adapter extension will recognize test inside `p1` and `p2`! You can even leave the config files inside the subfolders so that you can separately launch pytest in a single subfolder at a time (I've pushed a branch called `multiple_project` in the [demo repo](https://github.com/CarloDePieri/pytest-vscode-python-test-adapter-issue/tree/multiple_project) to demonstrate this setup).

**Comment by kondratyev-nv** [comment]
@CarloDePieri Thank you for the help!
@Makishima I'm closing this issue since the original problem was fixed. If you have any problems with discovering and running your tests feel free to create another issue and I'll try to help.
